documentation/API-GUIDELINES.md

```diff
@@ -15,20 +15,20 @@ The following paragraphs contain additional recommendations.
 
 ## Construction and Destruction of Drivers
 
-- Drivers should take peripherals via the `PeripheralRef` pattern - they don't consume peripherals directly.
-- If a driver requires pins, those pins should be configured using `fn with_signal_name(self, pin: impl Peripheral<P = InputConnection> + 'd) -> Self)` or `fn with_signal_name(self, pin: impl Peripheral<P = OutputConnection> + 'd) -> Self)`
+- Drivers must take peripherals via the `PeripheralRef` pattern - they don't consume peripherals directly.
+- If a driver requires pins, those pins must be configured using `fn with_signal_name(self, pin: impl Peripheral<P = InputConnection> + 'd) -> Self)` or `fn with_signal_name(self, pin: impl Peripheral<P = OutputConnection> + 'd) -> Self)`
 - If a driver supports multiple peripheral instances (for example, I2C0 is one such instance):
-  - The peripheral instance type should be positioned as the last type parameter of the driver type.
-  - The peripheral instance type should default to a type that supports any of the peripheral instances.
-  - The author is encouraged to use `crate::any_peripheral` to define the "any" peripheral instance type.
-  - The driver should implement a `new` constructor that automatically converts the peripheral instance into the any type, and a `new_typed` that preserves the peripheral type.
+  - The peripheral instance type must be positioned as the last type parameter of the driver type.
+  - The peripheral instance type must default to a type that supports any of the peripheral instances.
+  - The author must to use `crate::any_peripheral` to define the "any" peripheral instance type.
+  - The driver must implement a `new` constructor that automatically converts the peripheral instance into the any type, and a `new_typed` that preserves the peripheral type.
 - If a driver only supports a single peripheral instance, no instance type parameter is necessary.
```

Comment:
not necessary or not allowed?

Reply:
I'm wondering if we might want to revisit this line. I'm thinking of a case where a new chip comes out where it does have more than one of a peripheral that previously existed as a single peripheral in the past.

Maybe we should always enforce the type param, but instead recommend `Driver<T = SINGLE_INSTANCE>`, later down the line we can change this to `Driver<T = AnyInstance>`, in a non breaking way? I'm not actually sure if that would be breaking or not (provided we adjusted the constructor too.

Reply:
It might also be a reason to mandate type erasure so it's worth thinking about a little bit

Reply:
I've continued the discussion in https://github.com/esp-rs/esp-hal/issues/2572